src/ol/proj.js

```diff
@@ -467,6 +481,20 @@ export function getTransformFromProjections(
   if (!transformFunc) {
     transformFunc = identityTransform;
   }
+  let sourceMatrix = sourceProjection.getMatrix();
+  let destinationMatrix = destinationProjection.getMatrix();
+  if (sourceMatrix || destinationMatrix) {
+    const transform = transformFunc;
+    sourceMatrix = invertMatrix((sourceMatrix || createMatrix()).slice());
+    destinationMatrix = destinationMatrix || createMatrix();
+    const coordTransform = function (coordinate) {
+      return applyMatrix(
+        destinationMatrix,
+        transform(applyMatrix(sourceMatrix, coordinate.slice())),
+      );
+    };
+    transformFunc = createTransformFromCoordinateTransform(coordTransform);
+  }
```

Comment:
I wouldn't bake matrix transforms into projections. It does not feel right. Instead, a matrix transform should be defined with custom transform functions using `addCoordinateTransforms()`, like e.g. shown in https://openlayers.org/en/latest/examples/wms-custom-proj.html.

Reply:
The problem with that approach is transforms would need to be maintained between every matrix variation and every projection added (past and future) via proj4.  Given that this is a relatively unusual situation (although it could also be applied to KML ground overlays) I think setting an additional property on the projection and calculating the transforms dynamically as required would be easier.

Reply:
When you create a transform function, you can mix in the matrix transform part from a separate function, so you can still handle matrix variations outside the once created transform function. Basically the same thing you're currently doing in src/ol/proj.js.

Reply:
The existing transform functions are indexed by projection codes - the matrixes are assigned to projection objects with those codes.  Doing it this way avoids adding complexity to the indexing of the primary transforms.

Reply:
Can you point me to the code section where that happens?

Reply:
The new code calls `projection.setMatrix(matrix);` or `this.projection.setMatrix(matrix);` when determining the projection https://github.com/mike-000/openlayers/blob/7cc890b34747feb99173b974d2e6e143c88e0504/src/ol/source/GeoTIFF.js#L297
and if a matrix is detected when an override projection has been specified
https://github.com/mike-000/openlayers/blob/7cc890b34747feb99173b974d2e6e143c88e0504/src/ol/source/GeoTIFF.js#615

It is handled in existing reprojection code via `getTransform` and `transform` calls which need no changes as they are passed the projection objects (not codes) and therefore pass them on unchanged to `getTransformFromProjections`.